scripts/deprecate_release.ts

```diff
@@ -0,0 +1,220 @@
+}
+
+// check that the commit hash points to a release commit
+const { stdout: commitMessage } =
+  await $`git log -1 --pretty="%s" ${commitHash}`;
+if (!commitMessage.includes('Version Packages')) {
+  throw new Error(`
+    Expected commit message to include "Version Packages".
+    Instead found ${commitMessage}.
+    Make sure this commit hash points to a release commit.
+  `);
+}
+
+// check that this commit was made by the github-actions bot
+const { stdout: commitAuthor } =
+  await $`git log -1 --pretty="%an" ${commitHash}`;
+if (!commitAuthor.includes('github-actions[bot]')) {
+  throw new Error(`
+    Expected commit to be authored by github-actions[bot].
+    Instead found ${commitAuthor}.
+    Make sure this commit hash points to a release commit.
+  `);
+}
+
+// get the release tags associated with the commit
+// each tag name is the package name and version as '<packageName>@<version>'
+// this can be used directly as the input to npm deprecate
+const packageVersionsToDeprecate = await getTagsAtCommit(commitHash);
+
+if (packageVersionsToDeprecate.length === 0) {
+  throw new Error(`
+    Could not find any release tags associated with this commit.
+    Release deprecation is only supported for releases that have associated git tags.
+  `);
+}
+
+// get the most recent tag before the current release tag
+const { stdout: previousReleaseTag } =
+  await $`git describe ${packageVersionsToDeprecate[0]}^ --abbrev=0`;
+
+// get the commit hash associated with the previous release tag
+const { stdout: previousReleaseCommitHash } =
+  await $`git log -1 ${previousReleaseTag} --pretty=%H`;
+
+// get all the release tags associated with the previous commit
+// this is the list of package versions that need to be marked as latest
+const packageVersionsToRestore = await getTagsAtCommit(
+  previousReleaseCommitHash
+);
+if (packageVersionsToRestore.length === 0) {
+  throw new Error(`
+    Could not find any release tags for the previous release.
+    Ensure that there is previous release commit that is reachable from the current release commit in the git history.
+  `);
+}
+
+if (!(await isCleanWorkingTree())) {
+  throw new Error(`
+    Dirty working tree detected.
+    The release deprecation workflow requires a clean working tree to create the rollback PR.
+  `);
+}
+
+// now we need to grab this dist tag from the previous release tag
+// given a string like '@aws-amplify/auth-construct-alpha@0.6.0-beta.8', this regex will grab 'beta' from the string
+// for a non-pre release tag like @aws-amplify/backend-cli@0.10.0, this regex will not match and we default to the 'latest' tag
+// this logic assumes that all of the previous release package versions have the same dist tag
+// (this should always be the case when using changesets for releases)
+const distTagMatch = previousReleaseTag.match(
+  /\.\d+-(?<distTag>[\w\-.]+)\.\d+$/
+);
+
+const distTag = distTagMatch?.groups?.distTag ?? 'latest';
```

Comment:
would it help if workflow just took "previusReleaseCommitSha" as input parameter and get information from there ?

Reply:
nvm this. I re-read the code.

Reply:
finding the previous hash isn't really a problem. I think we could actually go the opposite direction and remove the commitHash input from the workflow entirely. Then the behavior would be
1. find the most recent release commit and deprecate those releases
2. find the second most recent release commit and point latest back to that

Reply:
This is a good idea. And we can just repeat workflow N times if we need to go back couple of releases.

Reply:
As written, repeating N times would not work. The script is updating metadata on npm, but it's not modifying git tags. So subsequent runs would deprecate the same release again

Reply:
Couple of options then.
1. Make inputs optional, if they're not provided look up commits.
2. Remove tags. 
3. While deprecating add additional tag to the commit (like `deprecated`) and skip that release next time if such tag is present.

Reply:
I think I like option 1 best, mostly because it requires the least amount of changes :)